### ISSUE TITLE
org.exist.util.ConfigurationHelper.getExistHome(String) generates wrong path for conf.xml

### ISSUE THREAD

**Issue by miljanuscumlic** [eXist-db/exist#1502]
### What is the problem
>  code throws java.nio.file.InvalidPathException when it is trying to find conf.xml file

```
14:51:43,843 ERROR [stderr] (default task-2) java.nio.file.InvalidPathException: Illegal char <:> at index 2: /C:/Program%20Files/wildfly-10.1.0.Final/bin/content/ruta-server-0.0.1.war/WEB-INF/classes/conf.xml
14:51:44,031 ERROR [stderr] (default task-2) 	at sun.nio.fs.WindowsPathParser.normalize(WindowsPathParser.java:182)
14:51:44,078 ERROR [stderr] (default task-2) 	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:153)
14:51:44,140 ERROR [stderr] (default task-2) 	at sun.nio.fs.WindowsPathParser.parse(WindowsPathParser.java:77)
14:51:44,187 ERROR [stderr] (default task-2) 	at sun.nio.fs.WindowsPath.parse(WindowsPath.java:94)
14:51:44,250 ERROR [stderr] (default task-2) 	at sun.nio.fs.WindowsFileSystem.getPath(WindowsFileSystem.java:255)
14:51:44,312 ERROR [stderr] (default task-2) 	at java.nio.file.Paths.get(Paths.java:84)
14:51:44,359 ERROR [stderr] (default task-2) 	at org.exist.util.ConfigurationHelper.getExistHome(ConfigurationHelper.java:97)
14:51:44,421 ERROR [stderr] (default task-2) 	at org.exist.util.ConfigurationHelper.getExistHome(ConfigurationHelper.java:32)
14:51:44,468 ERROR [stderr] (default task-2) 	at org.exist.util.ConfigurationHelper.lookup(ConfigurationHelper.java:139)
14:51:44,546 ERROR [stderr] (default task-2) 	at org.exist.util.ConfigurationHelper.lookup(ConfigurationHelper.java:116)
14:51:44,609 ERROR [stderr] (default task-2) 	at rs.ruta.server.CDR.registerParty(CDR.java:131)
14:51:44,671 ERROR [stderr] (default task-2) 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
14:51:44,765 ERROR [stderr] (default task-2) 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
14:51:44,828 ERROR [stderr] (default task-2) 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
14:51:44,890 ERROR [stderr] (default task-2) 	at java.lang.reflect.Method.invoke(Method.java:498)
14:51:44,937 ERROR [stderr] (default task-2) 	at org.jboss.as.ee.component.ManagedReferenceMethodInterceptor.processInvocation(ManagedReferenceMethodInterceptor.java:52)
14:51:45,000 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,046 ERROR [stderr] (default task-2) 	at org.jboss.as.ee.concurrent.ConcurrentContextInterceptor.processInvocation(ConcurrentContextInterceptor.java:45)
14:51:45,109 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,171 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InitialInterceptor.processInvocation(InitialInterceptor.java:21)
14:51:45,234 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,281 ERROR [stderr] (default task-2) 	at org.jboss.invocation.ChainedInterceptor.processInvocation(ChainedInterceptor.java:61)
14:51:45,343 ERROR [stderr] (default task-2) 	at org.jboss.as.ee.component.interceptors.ComponentDispatcherInterceptor.processInvocation(ComponentDispatcherInterceptor.java:52)
14:51:45,406 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,453 ERROR [stderr] (default task-2) 	at org.jboss.as.webservices.deployers.WSComponentInstanceAssociationInterceptor.processInvocation(WSComponentInstanceAssociationInterceptor.java:56)
14:51:45,500 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,578 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.run(InterceptorContext.java:356)
14:51:45,640 ERROR [stderr] (default task-2) 	at org.wildfly.security.manager.WildFlySecurityManager.doChecked(WildFlySecurityManager.java:636)
14:51:45,703 ERROR [stderr] (default task-2) 	at org.jboss.invocation.AccessCheckingInterceptor.processInvocation(AccessCheckingInterceptor.java:61)
14:51:45,765 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,828 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.run(InterceptorContext.java:356)
14:51:45,875 ERROR [stderr] (default task-2) 	at org.jboss.invocation.PrivilegedWithCombinerInterceptor.processInvocation(PrivilegedWithCombinerInterceptor.java:80)
14:51:45,921 ERROR [stderr] (default task-2) 	at org.jboss.invocation.InterceptorContext.proceed(InterceptorContext.java:340)
14:51:45,984 ERROR [stderr] (default task-2) 	at org.jboss.invocation.ChainedInterceptor.processInvocation(ChainedInterceptor.java:61)
14:51:46,031 ERROR [stderr] (default task-2) 	at org.jboss.as.ee.component.ViewService$View.invoke(ViewService.java:198)
14:51:46,093 ERROR [stderr] (default task-2) 	at org.jboss.as.webservices.invocation.AbstractInvocationHandler.invoke(AbstractInvocationHandler.java:137)
14:51:46,171 ERROR [stderr] (default task-2) 	at org.jboss.wsf.stack.cxf.JBossWSInvoker.performInvocation(JBossWSInvoker.java:169)
14:51:46,218 ERROR [stderr] (default task-2) 	at org.apache.cxf.service.invoker.AbstractInvoker.invoke(AbstractInvoker.java:96)
14:51:46,281 ERROR [stderr] (default task-2) 	at org.apache.cxf.jaxws.AbstractJAXWSMethodInvoker.invoke(AbstractJAXWSMethodInvoker.java:232)
14:51:46,343 ERROR [stderr] (default task-2) 	at org.apache.cxf.jaxws.JAXWSMethodInvoker.invoke(JAXWSMethodInvoker.java:85)
14:51:46,406 ERROR [stderr] (default task-2) 	at org.jboss.wsf.stack.cxf.JBossWSInvoker.invoke(JBossWSInvoker.java:145)
14:51:46,453 ERROR [stderr] (default task-2) 	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$1.run(ServiceInvokerInterceptor.java:59)
14:51:46,515 ERROR [stderr] (default task-2) 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
14:51:46,578 ERROR [stderr] (default task-2) 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
14:51:46,625 ERROR [stderr] (default task-2) 	at org.apache.cxf.interceptor.ServiceInvokerInterceptor$2.run(ServiceInvokerInterceptor.java:126)
14:51:46,671 ERROR [stderr] (default task-2) 	at org.apache.cxf.workqueue.SynchronousExecutor.execute(SynchronousExecutor.java:37)
14:51:46,734 ERROR [stderr] (default task-2) 	at org.apache.cxf.interceptor.ServiceInvokerInterceptor.handleMessage(ServiceInvokerInterceptor.java:131)
14:51:46,781 ERROR [stderr] (default task-2) 	at org.apache.cxf.phase.PhaseInterceptorChain.doIntercept(PhaseInterceptorChain.java:308)
14:51:46,843 ERROR [stderr] (default task-2) 	at org.apache.cxf.transport.ChainInitiationObserver.onMessage(ChainInitiationObserver.java:121)
14:51:46,906 ERROR [stderr] (default task-2) 	at org.apache.cxf.transport.http.AbstractHTTPDestination.invoke(AbstractHTTPDestination.java:254)
14:51:46,953 ERROR [stderr] (default task-2) 	at org.jboss.wsf.stack.cxf.RequestHandlerImpl.handleHttpRequest(RequestHandlerImpl.java:108)
14:51:47,015 ERROR [stderr] (default task-2) 	at org.jboss.wsf.stack.cxf.transport.ServletHelper.callRequestHandler(ServletHelper.java:134)
14:51:47,062 ERROR [stderr] (default task-2) 	at org.jboss.wsf.stack.cxf.CXFServletExt.invoke(CXFServletExt.java:88)
14:51:47,125 ERROR [stderr] (default task-2) 	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.handleRequest(AbstractHTTPServlet.java:298)
14:51:47,187 ERROR [stderr] (default task-2) 	at org.apache.cxf.transport.servlet.AbstractHTTPServlet.doPost(AbstractHTTPServlet.java:217)
14:51:47,234 ERROR [stderr] (default task-2) 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
14:51:47,296 ERROR [stderr] (default task-2) 	at org.jboss.wsf.stack.cxf.CXFServletExt.service(CXFServletExt.java:136)
14:51:47,343 ERROR [stderr] (default task-2) 	at org.jboss.wsf.spi.deployment.WSFServlet.service(WSFServlet.java:140)
14:51:47,453 ERROR [stderr] (default task-2) 	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
14:51:47,515 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletHandler.handleRequest(ServletHandler.java:85)
14:51:47,578 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.security.ServletSecurityRoleHandler.handleRequest(ServletSecurityRoleHandler.java:62)
14:51:47,640 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletDispatchingHandler.handleRequest(ServletDispatchingHandler.java:36)
14:51:47,687 ERROR [stderr] (default task-2) 	at org.wildfly.extension.undertow.security.SecurityContextAssociationHandler.handleRequest(SecurityContextAssociationHandler.java:78)
14:51:47,734 ERROR [stderr] (default task-2) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
14:51:47,796 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.security.SSLInformationAssociationHandler.handleRequest(SSLInformationAssociationHandler.java:131)
14:51:47,843 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.security.ServletAuthenticationCallHandler.handleRequest(ServletAuthenticationCallHandler.java:57)
14:51:47,906 ERROR [stderr] (default task-2) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
14:51:47,953 ERROR [stderr] (default task-2) 	at io.undertow.security.handlers.AbstractConfidentialityHandler.handleRequest(AbstractConfidentialityHandler.java:46)
14:51:48,015 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.security.ServletConfidentialityConstraintHandler.handleRequest(ServletConfidentialityConstraintHandler.java:64)
14:51:48,062 ERROR [stderr] (default task-2) 	at io.undertow.security.handlers.AuthenticationMechanismsHandler.handleRequest(AuthenticationMechanismsHandler.java:60)
14:51:48,125 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.security.CachedAuthenticatedSessionHandler.handleRequest(CachedAuthenticatedSessionHandler.java:77)
14:51:48,187 ERROR [stderr] (default task-2) 	at io.undertow.security.handlers.NotificationReceiverHandler.handleRequest(NotificationReceiverHandler.java:50)
14:51:48,234 ERROR [stderr] (default task-2) 	at io.undertow.security.handlers.AbstractSecurityContextAssociationHandler.handleRequest(AbstractSecurityContextAssociationHandler.java:43)
14:51:48,296 ERROR [stderr] (default task-2) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
14:51:48,343 ERROR [stderr] (default task-2) 	at org.wildfly.extension.undertow.security.jacc.JACCContextIdHandler.handleRequest(JACCContextIdHandler.java:61)
14:51:48,390 ERROR [stderr] (default task-2) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
14:51:48,453 ERROR [stderr] (default task-2) 	at io.undertow.server.handlers.PredicateHandler.handleRequest(PredicateHandler.java:43)
14:51:48,500 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler.handleFirstRequest(ServletInitialHandler.java:292)
14:51:48,562 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler.access$100(ServletInitialHandler.java:81)
14:51:48,625 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:138)
14:51:48,671 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler$2.call(ServletInitialHandler.java:135)
14:51:48,718 ERROR [stderr] (default task-2) 	at io.undertow.servlet.core.ServletRequestContextThreadSetupAction$1.call(ServletRequestContextThreadSetupAction.java:48)
14:51:48,781 ERROR [stderr] (default task-2) 	at io.undertow.servlet.core.ContextClassLoaderSetupAction$1.call(ContextClassLoaderSetupAction.java:43)
14:51:48,828 ERROR [stderr] (default task-2) 	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
14:51:48,890 ERROR [stderr] (default task-2) 	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
14:51:48,937 ERROR [stderr] (default task-2) 	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
14:51:49,000 ERROR [stderr] (default task-2) 	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
14:51:49,062 ERROR [stderr] (default task-2) 	at io.undertow.servlet.api.LegacyThreadSetupActionWrapper$1.call(LegacyThreadSetupActionWrapper.java:44)
14:51:49,109 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler.dispatchRequest(ServletInitialHandler.java:272)
14:51:49,171 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler.access$000(ServletInitialHandler.java:81)
14:51:49,234 ERROR [stderr] (default task-2) 	at io.undertow.servlet.handlers.ServletInitialHandler$1.handleRequest(ServletInitialHandler.java:104)
14:51:49,281 ERROR [stderr] (default task-2) 	at io.undertow.server.Connectors.executeRootHandler(Connectors.java:202)
14:51:49,328 ERROR [stderr] (default task-2) 	at io.undertow.server.HttpServerExchange$1.run(HttpServerExchange.java:805)
14:51:49,390 ERROR [stderr] (default task-2) 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
14:51:49,437 ERROR [stderr] (default task-2) 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
14:51:49,484 ERROR [stderr] (default task-2) 	at java.lang.Thread.run(Thread.java:748)
```

### What did you expect
> Not to throw exception

### Describe how to reproduce or add a test
Lots of things from the list are not that important but I am putting them all for completeness
> - Deploy exist-db as a web application in a Jetty servlet container at http:://localhost:8888/exist/xmlrpc
> - Create Java web application as mavenized project (with Eclipse Oxygen)
> - Create javax.jws.WebService that has javax.jws.WebMethod with the line 
Path configPath = ConfigurationHelper.lookup("conf.xml");
> - Put conf.xml inside src/main/resources directory
> - Deploy java web application on the WildFly 10.1.0 application server.
> - Create client application
> - Call WebService from the client

### Context information
- eXist-db version eXist-db 3.4.0
- Java8u141
- Windows XP
- 32
- Any custom changes in e.g. conf.xml: data directory changed

I think that this is a bug and that it could possibly be fixed if the code from the method org.exist.util.ConfigurationHelper.getExistHome(String)
```
 // try classpath
 final URL configUrl = ConfigurationHelper.class.getClassLoader().getResource(config);
        if (configUrl != null) {
            final Path existHome = Paths.get(configUrl.getPath()).getParent();
            LOG.debug("Got eXist home from classpath: {}", existHome.toAbsolutePath().toString());
            return Optional.of(existHome);
        }
```

is replaced with something like this:

```
 // try classpath
        final URL configUrl = ConfigurationHelper.class.getClassLoader().getResource(config);
        if (configUrl != null) {
         	final String configPath = configUrl.getPath().replaceFirst("^/(.:/)", "$1");
        	final Path existHome = Paths.get(configPath).getParent();
            LOG.debug("Got eXist home from classpath: {}", existHome.toAbsolutePath().toString());
            return Optional.of(existHome);
        }
```
I have not tested this code, and don't know if it conflicts with other use cases. And I am completely newbie :)

Explanation:
Stacktrace is pretty much self explanatory, especially for you guys.
Object configUrl has a field `path` which in my case has prepended one character "/" : `/C:/Program%20Files/wildfly-10.1.0.Final/bin/content/ruta-server-0.0.1.war/WEB-INF/classes/conf.xml`. Why is that I don't know, but it should be removed, because that `String` is passed to the `Paths.get(String)` method which then goes up to `sun.nio.fs.WindowsPathParser.normalize` which eventually throws the exception because the path is not valid. Here is the screenshot from my debugging session:
![conf2](https://user-images.githubusercontent.com/30440470/28728813-589f711a-73ca-11e7-868b-6f72491a810b.jpg)

**Comment by dizzzz** [comment]
Please could you verify that wildfly does 'explode' the warfile on the disk ? (please note: you do not mention explicitly you try to deploy the war file in this container....)

please note that we do not recommend eXist-db to be deployed as  a war file. It provides several risks, and no WAR container behaves identical.

The users that deploy an eXist-db WAR file, do this in Tomcat. In tomcat things work reasonably well, from what I see in reports....

**Comment by miljanuscumlic** [comment]
I will answer on the first couple of questions and on the other ones some other time.

My intention is to deploy exist-db as war file in the Wildfly container but at this point this is not the case. Exist-db is on its own jetty server. My web service (server) application ruta.war is deployed on Wildfly. From that application I am calling exist-db.

I think that the exception that is thrown is not in the relation to the fact that ruta.war is extracted or not. As I see it, the point is that getResource() method in `final URL configUrl = ConfigurationHelper.class.getClassLoader().getResource(config);` returns URL object that has not valid path in itself because of the first character in it which is `/`. And that's why the exception is thrown whether there is conf.xml actually on that path or not. The exception would be thrown and when the file is on that path, I believe.

See similar issues:
https://stackoverflow.com/questions/43972777/exception-in-thread-main-java-nio-file-invalidpathexception-illegal-char
https://stackoverflow.com/questions/9834776/java-nio-file-path-issue

**Comment by dizzzz** [comment]
It is not possible to call Java an eXist-db JAVA API (running on server 1) from server 2;
The best thing to do is to send HTTP requests from the Wildfly app to eXist's REST/RESTXQ interface.

Second best is to use the XMLDB Java API to perform remote operations.

https://youtu.be/_0CylabK-VQ
See http://exist-db.org/exist/apps/doc/devguide_xmldb.xml?q=XMLDB&field=all&id=D2.2.3#D2.2.3

**Comment by dizzzz** [comment]
Please switch to exist-open for questions/support ; we'd like to keep github to contain concrete issues. (It helps to refer to this item, of course)

**Comment by miljanuscumlic** [comment]
Sorry because I was not clear enough, but I am using XMLDB API, and my intention was not to search for support from you. Thanks for investing your time in this exception and matter.

**Comment by dizzzz** [comment]
well I am happy to provide support and provide answers :-) The mailinglist is however often a much more efficient medium (more ppl read the posts)

**Comment by dizzzz** [comment]
from the stack trace, i don't see that the (remote) XMLDB classes are used. For a remote connection, I don't think `org.exist.util.ConfigurationHelper` needs to be used. If it is the case, we have a problem :-)

please check for inspiration https://github.com/eXist-db/exist/blob/develop/samples/src/org/exist/examples/xmldb/Put.java

**Comment by miljanuscumlic** [comment]
Hi dizzzz,
you are right about me using wrong API. I am using XMLDB but I mixed it up with eXist-db JAVA API which is the product of my ignorance. So, there is no problem in regard to this.
Thanks for the inspiration you posted, I have studied it before and lots of other things. In my code I wanted to do more complex things (i.e. user management) but I won't bother you with that. You have pointed me correctly to the mailing list. Thanks again.
